File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.py
class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        n = len(s)
        dp = [0] * (n+1)
        for i in range(1, n+1):
            dp[i] = 1 + dp[i-1]
            for j in range(0, i):
                if s[j:i:] in dictionary:
                    dp[i] = min(dp[i], dp[j])
        return dp[n]

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.js
/**
 * @param {string} s
 * @param {string[]} dictionary
 * @return {number}
 */
var minExtraChar = function(s, dictionary) {
    const n = s.length
    dictionary = new Set(dictionary)
    const dp = Array(n+1).fill(Infinity)
    dp[0] = 0
    for (let i=1; i <= n; i++) {
        dp[i] = 1 + dp[i-1]
        for (let j=0; j < i; j++) {
            if (dictionary.has(s.substring(j, i))) {
                dp[i] = Math.min(dp[i], dp[j])
            }
        }
    }
    return dp[n]
};